### PR TITLE
Zephyr: re-work DT flash partitioning for Nordic nRF boards in the tree

### PR DESCRIPTION
**Pull request by ioannisg**
The purpose of this PR is to address #27471 which is a rather serious bug: the ARM linker is not aware of the storage partition, when the flash partitioning layout - defined in board DeviceTree - is not used by the linker. If the Zephyr firmware is relatively large, it may overflow into the storage partition, with unpredictable behavior.

To solve the problem, we need to always use the Device Tree-defined partition layout information for the Zephyr application code partition, in the cases when there are parts of flash (e.g. the storage partition) that cannot be available for code. So we need to enforce USE_DT_CODE_PARTITION at all times.

This PR shows how this can be done and actually applies this for the Nordic-maintained nRF-based platforms in the tree. 

This PR is an alternative to #29749. It is aligned with the constructive feedback by @nvlsianpu and @Laczen .

The main design idea is that most platforms would need to have **two alternative flash partitioning layouts to be used when building with or without bootloader, respectively.** The generic logic to pick the right layout is done in Kconfig.zephyr. Platforms where more alternative layouts make sense would need to define a more complex logic (in the board definition)

In brief, this PR - consisting of several patches - does the following:
1. Adds a dependency to `config USE_DT_CODE_PARTITION`, so it cannot be selected if the device tree does not define the corresponding "chosen" `code-partition` symbols.
2. Instructs the linker to use the `zephyr,code-partition` when building a chain-loadable image (loadable by MCUboot) and `zephyr,single-image-code-partition` when building a standalone image . How?  Kconfig.zephyr will modify the defaults for FLASH_LOAD_OFFSET and FLASH_LOAD_SIZE, to pick either the single-image-code-partition, when we build a standalone image, or pick code-partition, when we build a chain-loadable image (with MCUboot support).
4. For all Nordic nRF based boards, the PR defines two alternative flash layouts (with and without MCUboot) and defines the chosen symbols, for code partitions with and without BL. The PR forces using the DT Code Partition by default in Nordic DKs, regardless of MCUBoot, **effectively solving #27471 for them**.
5. For Trustzone-enabled Nordic platforms (nrf5340 APP and nRF9160) the flash layouts also depend on the security domain we build for (Secure or Non-Secure), therefore the default mechanism we implement in Kconfig.zephyr is overridden locally in the nrf53 and nrf91 board definitions.

### CODE REVIEW

**Comment by ioannisg** [comment]
Currently address only for Nordic-maintained nRF Dev Kits, but if this solution is approved, it should be extended to all boards.

**Review by nvlsianpu — COMMENTED** [review]
Intoduce duplicated DT node is not a possible solution - need to use overlays

File: boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts [hunk 1]
```diff
@@ -261,16 +263,22 @@ arduino_spi: &spi3 {
...
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
 
-		/*
-		 * Storage partition will be used by FCB/LittleFS/NVS
-		 * if enabled.
-		 */
+		code_partition: partition@0 {
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:278]: I would not favor overlays - defining both alternative layouts in the board definitions is a cleaner solution. To overcome the problem you spotted, we could alter the node name prefix or suffix for the boot partition, so it removes this compilation issue.

**Comment by ioannisg** [comment]
@Laczen @galak @mbolivar-nordic @anangl I am looking, again, for feedback on the rework of DT flash layout/partitioning.

File: boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ioannisg) [reply boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:268]: @nvlsianpu pls, take a look at the alternative sketching of the layouts now.

**Comment by Laczen** [comment]
@ioannisg, it would be the cleanest using overlays, I think it is "bending the rules" a bit to declare two partitions at the same starting address. There could be the following options:
a. No code-partition specified: complete flash area is zephyr-code (no storage possibility)
b. Code partition specified and determined: possibility to use storage if defined
c. Mcuboot layout with code-partition = boot_partition: Offset = 0 and no extra offset.
d. Mcuboot layout with code-partition = slot0_partition: Offset = slot0_offset + image_offset (0x200)

**Comment by ioannisg** [comment]
> @ioannisg, it would be the cleanest using overlays, I think it is "bending the rules" a bit to declare two partitions at the same starting address. There could be the following options:
> a. No code-partition specified: complete flash area is zephyr-code (no storage possibility)
> b. Code partition specified and determined: possibility to use storage if defined
> c. Mcuboot layout with code-partition = boot_partition: Offset = 0 and no extra offset.
> d. Mcuboot layout with code-partition = slot0_partition: Offset = slot0_offset + image_offset (0x200)

@Laczen to clarify: you suggest having a dts overlay inside every board definition?

**Comment by nvlsianpu** [comment]
> it would be the cleanest using overlays, I think it is "bending the rules" a bit to declare two partitions at the same starting address.

This silent rule is indeed not satisfy by this approach, but it makes usage simpler for an end-user as it doesn't requires usage of overlays. Can we neglect this rule then? Moreover It was never stated firmly.

File: boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts:268]: looks good, I can accept such solution

**Comment by ioannisg** [comment]
> > it would be the cleanest using overlays, I think it is "bending the rules" a bit to declare two partitions at the same starting address.
> 
> This silent rule is indeed not satisfy by this approach, but it makes usage simpler for an end-user as it doesn't requires usage of overlays. Can we neglect this rule then? Moreover It was never stated firmly.

I am not aware of such a formal rule, either. I understand the "beauty" of having a partitions' definition that is cleaner, but, using a .dts overlay for MCUboot, for each board definition seems an overkill; i think we could go for the "hybrid" solution of having a single partitions DTS node with possibly multiple layouts inside it. And proper comments to make it easy to understand. 

Also, overlays means we duplicate partition entries (e.g. storage) which will have to be defined in all the different layouts (the main one and the overlay or overlays). Currently, this is not needed since all partitions are defined in one place..

**Comment by ioannisg** [comment]
Thinking a bit more, @Laczen, I am not sure how the overlay could work: the overlay is already effective during DTS compilation; at that point we do not know if we build a standalone or a chain-loadable image. So we need to have both alternative layouts at our disposal; an overlay would simply overwrite the default layout.

[In MCUboot we create and use a partition overlay, but we know already at DT compilation time that we will be building a bootloader.]

Am I missing something? feel free to draft a proposal.

**Comment by Laczen** [comment]
> Thinking a bit more, @Laczen, I am not sure how the overlay could work: the overlay is already effective during DTS compilation; at that point we do not know if we build a standalone or a chain-loadable image. So we need to have both alternative layouts at our disposal; an overlay would simply overwrite the default layout.
> 
> [In MCUboot we create and use a partition overlay, but we know already at DT compilation time that we will be building a bootloader.]
> 
> Am I missing something? feel free to draft a proposal.

This can be solved by deleting the default layout.

Regarding the default definition: Be carefull not to give mixed signals: "you cannot define overlapping regions" (because you will get stuck) and "we will define overlapping regions (because it is easier)".

**Comment by ioannisg** [comment]
> --This can be solved by redefining the chosen partition.-- I think I misunderstood your remark. A standalone image has a different flash layout than chain-loadable, so both should have their own layout.
 
@Laczen no worries. Yes, I think this the middle ground here; we need to have the different layouts available.

> Regarding the default definition: Be carefull not to give mixed signals: "you cannot define overlapping regions" (because you will get stuck) and "we will define overlapping regions (because it is easier)".

Agreed. I am thinking of a "big" warning that
- we are allowed to have multiple flash layouts (because we need to have them at build time before Kconfig)
- but we cannot use partitions from mixed layouts as we risk messing up with the linker.

And add this warning in the .dts files of the flash layout definitions.

**Comment by ioannisg** [comment]
> > --This can be solved by redefining the chosen partition.-- I think I misunderstood your remark. A standalone image has a different flash layout than chain-loadable, so both should have their own layout.
> 
> @Laczen no worries. Yes, I think this the middle ground here; we need to have the different layouts available, if we want to follow the nice approach of always using the DTS partitioning when possible (and at least, when we do dangerous things like using storage partition)

> 
> > Regarding the default definition: Be carefull not to give mixed signals: "you cannot define overlapping regions" (because you will get stuck) and "we will define overlapping regions (because it is easier)".
> 
> Agreed. I am thinking of a "big" warning that
> 
> * we are allowed to have multiple flash layouts (because we need to have them at build time before Kconfig)
> * but we cannot use partitions from mixed layouts as we risk messing up with the linker.
> 
> And add this warning in the .dts files of the flash layout definitions.

**Comment by de-nordic** [comment]
I have tried to solve this, from different reasons, in several other ways and I think that this solution is fine.

**Review by nvlsianpu — COMMENTED** [review]
looks fine, although have to spend some more time on checking that

**Comment by ioannisg** [comment]
BTW, related to this PR, I opened this MCUboot issue: https://github.com/mcu-tools/mcuboot/issues/915

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts [hunk 2]
```diff
@@ -231,12 +232,18 @@ fem_spi: &spi3 {
...
+	 * Note: A Zephyr build should use partitions belonging
+	 * to the same flash layout.
+	 */
 	partitions {
 		compatible = "fixed-partitions";
 		#address-cells = <1>;
 		#size-cells = <1>;
 
-		boot_partition: partition@0 {
+	/* Flash layout with MCUboot */
```
Comment (mbolivar-nordic) [inline boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:245]: Really not sure if overlapping partitions is a good idea here. Curious what @galak and @pabigot think.

File: boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig [hunk 3]
```diff
@@ -59,26 +59,53 @@ endif # BUILD_WITH_TFM
...
 
-if (BOARD_NRF5340PDK_NRF5340_CPUAPP || BOARD_NRF5340DK_NRF5340_CPUAPP) && TRUSTED_EXECUTION_SECURE
+if (BOARD_NRF5340PDK_NRF5340_CPUAPP || BOARD_NRF5340DK_NRF5340_CPUAPP)
+
+config HAS_DT_CODE_PARTITION
+	bool
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_CODE_S_PARTITION)) if TRUSTED_EXECUTION_SECURE && BOOTLOADER_MCUBOOT
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_SINGLE_IMAGE_CODE_S_PARTITION)) if TRUSTED_EXECUTION_SECURE
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_CODE_PARTITION)) if TRUSTED_EXECUTION_SECURE && BOOTLOADER_MCUBOOT
+	default $(dt_chosen_enabled,$(DT_CHOSEN_Z_SINGLE_IMAGE_CODE_PARTITION)) if TRUSTED_EXECUTION_SECURE
```
Comment (mbolivar-nordic) [inline boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig:76]: The number of cases that are influencing HAS_DT_CODE_PARTITION is starting to explode. Not sure about this either.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I personally think "the cure is worse than the disease" here. Yes, users might misconfigure their partitions and end up with text overlapping with settings, but I'm not sure that all this extra machinery is a good solution. Would like to discuss in dev review.

**Comment by Laczen** [comment]
> I personally think "the cure is worse than the disease" here. Yes, users might misconfigure their partitions and end up with text overlapping with settings, but I'm not sure that all this extra machinery is a good solution. Would like to discuss in dev review.

I agree with @mbolivar-nordic: "The cure is worse than the disease".

I propose the following alternative solution:
a. A new standard partitioning is created for boards that are using the storage area, this consists of a area called slot0_partition and a storage area (for devices that have secure, non-secure areas the slot0_partition could be divided).
b. The code-partition is standard set to slot0_partition (for all devices).
c. A mcuboot overlay is added to the boards directory, this mcuboot overlay deletes the slot0_partition and redefines the partitions.
d. Projects using mcuboot use the overlay to generate the image.
e. The mcuboot project itself uses a overlay that sets the code-partition to mcuboot.

The advantages:
a. It is clear and uses the available options as they were meant to be used.
b. It separates mcuboot partitions from default partitions, opening the way for alternative bootloaders.
c. It creates no special meaning for the storage partition.

The disadvantage: more work (altough maybe not).

A possible risk is that by separating the mcuboot partitioning to a different file it is easier to overwrite the storage partition.

File: Kconfig.zephyr [hunk 4]
```diff
@@ -80,8 +80,26 @@ config HAS_FLASH_LOAD_OFFSET
 
 if HAS_FLASH_LOAD_OFFSET
 
+config HAS_DT_CODE_PARTITION
+	bool
+	default y if MCUBOOT
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply Kconfig.zephyr:90]: The reason for this is that mcuboot currently adds a dts .overlay, choosing the boot partition as the code partition (which in my opinion should **not** be done in mcuboot but in zephyr; i've opened https://github.com/mcu-tools/mcuboot/issues/915 for that). When this is moved to Zephyr, we can address the point you mention here correctly, but for now I wanted this PR to go through regardless of the progress of the mcuboot fix.
Note that this `default y if MCUBOOT` does not make things worse that they already are: `config MCUBOOT` already selects USE_DT_CODE_PARTITION unconditionally, and does not check whether the DT flash partition definition actually exists.

File: boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts (hunk 2 above)
Comment (ioannisg) [reply boards/arm/nrf21540dk_nrf52840/nrf21540dk_nrf52840.dts:245]: I've added a documentation comment - everywhere where the flash partitioning is defined - stating that the user shall always restrict to use partitions belonging to the same flash layout, i.e. not to use overlapping partitions.

File: boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig (hunk 3 above)
Comment (ioannisg) [reply boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig:76]: This is a special case for TrustZone-enable devices that want to support both running standalone Secure-only execution environments and Secure/Non-Secure environments.

File: Kconfig.zephyr [hunk 5]
```diff
@@ -80,9 +81,15 @@ config HAS_FLASH_LOAD_OFFSET
 
 if HAS_FLASH_LOAD_OFFSET
 
+# Workaround for not being able to have commas in macro arguments
+DT_CHOSEN_Z_CODE_PARTITION := zephyr,code-partition
+DT_CHOSEN_Z_SINGLE_IMAGE_CODE_PARTITION := zephyr,single-image-code-partition
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply Kconfig.zephyr:86]: > In short, the question seems to be "are we using any partitions?". If "yes", we need to link into a code partition. If "no", the text section can take all of flash.

I could agree on not enforcing USE_DT_CODE_PARTITION by default, if we come up with a mechanism that lets the build system know whether we use _any_ code partitions. 

> Whether we're a "single" image or not doesn't really seem like the right question.

The problem is that the code-partition boundaries should different when we use bootloader and when we do not. It cannot be the same code partition boundaries. Whether we use bootloader or not is something we learn at Kconfig stage, while the partitions need to be defined, earlier, at DT stage.

**Review by galak — CHANGES_REQUESTED** [review]
I'm opposed to adding a new devicetree property.  I'd rather see this config choice specified via cmake somehow.

**Pull request closed without merging**